### PR TITLE
fix kdenlive read if global_feed=1 is missing

### LINKED ISSUE #1159: Kdenlive adapter no longer working
### Incorrect Functionality and General Questions

The kdenlive adapter does not seem to work any more. I looked into the python code a bit and it looks like it's expecting the kdenlive project file to look completely different than it does, it's likely that kdenfile has changed their file format and how they handle tracks in it quite significantly.

```
~/Projekte/sonnenblumenvideo/ otioconvert -i sonnenblumenvideo.kdenlive -o sonnenblumenvideo.fcpxml
Traceback (most recent call last):
  File "/usr/bin/otioconvert", line 33, in <module>
    sys.exit(load_entry_point('OpenTimelineIO==0.14.0', 'console_scripts', 'otioconvert')())
  File "/usr/lib/python3.9/site-packages/opentimelineio/console/otioconvert.py", line 263, in main
    result_tl = otio.adapters.read_from_file(
  File "/usr/lib/python3.9/site-packages/opentimelineio/adapters/__init__.py", line 143, in read_from_file
    return adapter.read_from_file(
  File "/usr/lib/python3.9/site-packages/opentimelineio/adapters/adapter.py", line 139, in read_from_file
    result = self._execute_function(
  File "/usr/lib/python3.9/site-packages/opentimelineio/plugins/python_plugin.py", line 174, in _execute_function
    return (getattr(self.module(), func_name)(**kwargs))
  File "/usr/lib/python3.9/site-packages/opentimelineio_contrib/adapters/kdenlive.py", line 95, in read_from_string
    for maintrack in maintractor.findall('track'):
AttributeError: 'NoneType' object has no attribute 'findall'
```

This problem has also been described here: https://www.mail-archive.com/kde-bugs-dist@kde.org/msg620606.html

## To Reproduce

1. Operating System: Arch LInux – likely doesn't matter
2. Python version: 3.9 - likely doesn't matter
3. OpenTimelineIO release version: 0.14.0
4. Create a project with the newest version of kdenlive (21.08.3 in my case). put in a video or something
5. use the opentimelineio export feature or use otioconvert to try to convert it, get the error message

## Expected Behavior

Should be able to read a kdenlive project file.

## More information

This is probably helpful: https://community.kde.org/Kdenlive/Development/File_format

### PR DESCRIPTION
**Pull request by codingcatgirl**
In newer versions of kdenlive, the global_feed=1 attribute isn't set in the "main" tractor.
However, it looks like the last tractor in the xml is still the main tractor.
This change uses the last tractor as the main tractor if it can't be found using global_feed=1 and, just to be safe, checks if all other tractors are included by it.

This fixes #1159 as well as the following bugs in the KDE bugtracker:
https://bugs.kde.org/show_bug.cgi?id=426942
https://bugs.kde.org/show_bug.cgi?id=445214

### CODE REVIEW

**Comment by scanlime** [comment]
Thank you for the patch, works well here!

**Comment by evilscientress** [comment]
Now I'm curious. @scanlime if I remember correctly you also use Premiere, @codingcatgirl and I had the issue that when we transferred a kdenlive project to Premiere 2021 via fxp7 xml the timelines all had the wrong resolution (NTSC SD instead of 1080p30) and the meta data of video files got corrupted. They all had 2 mono audio tracks instead of 1 stereo audio track.

Did you experience similar issues?

**Comment by scanlime** [comment]
> Now I'm curious. @scanlime if I remember correctly you also use Premiere, @codingcatgirl and I had the issue that when we transferred a kdenlive project to Premiere 2021 via fxp7 xml the timelines all had the wrong resolution (NTSC SD instead of 1080p30) and the meta data of video files got corrupted. They all had 2 mono audio tracks instead of 1 stereo audio track.
> 
> Did you experience similar issues?

Sorry, I don't have Premiere any more. Been trying to build a new workflow around kdenlive and I've been using otio to write some scripts for that.

**Review by gilou — APPROVED** [review]
This fixes it for me, otio can then properly read my recent .kdenlive file.

**Review by meshula — APPROVED** [review]
minor typo: lets => let's

Thanks for the fix

**Comment by jminor** [comment]
Thanks @codingcatgirl - could you submit a Contributor License Agreement as outlined here? https://opentimelineio.readthedocs.io/en/latest/tutorials/contributing.html

**Comment by codingcatgirl** [comment]
@jminor i will probably do so once i figure out how to make my printer work again, but first i need to think about whether i feel okay filling that form out with my address and phone number. 

for the record i don't mind if you fix the bug some other way

**Comment by jminor** [comment]
@codingcatgirl our project requires a CLA from all contributors, but we completely understand that some people might not feel comfortable signing it for a wide variety of reasons. If it helps, we do keep the signed CLAs carefully protected and only visible to our legal department (at Pixar and soon at the Academy Software Foundation) and our project's Technical Steering Committee. That's a total of about 8-10 people.
Also, if your printer is not working, we have accepted digital signatures before (e.g. from a PDF reader/annotator, or just an image editing software).
I hope that helps :)

**Comment by jminor** [comment]
Oh, and to be clear the only thing that is publicly shared is the code, and your github username.
You can also choose to be included in the CONTRIBUTORS.md file if you want.

**Comment by scanlime** [comment]
> Oh, and to be clear the only thing that is publicly shared is the code, and your github username. You can also choose to be included in the CONTRIBUTORS.md file if you want.

Can you fix the bug without soliciting personal information from someone who's already gone far out of their way to help? This makes the project look ridiculous.

**Comment by jminor** [comment]
Hi @codingcatgirl and @scanlime - we discussed the concerns you raised amongst our steering committee and asked the legal advisors for this project if there is any way to accept contributions without contact information. At the moment the answer is no, we are not allowed to. However, we do see the importance of lowering the barriers for anyone to contribute from an inclusion & diversity perspective, and a general open-source community perspective, and we would like to be as inclusive as we can and so we will continue to look for ways to address this concern in the future.

Here are some options for moving forward with this PR:
- The author of this PR could choose to sign the CLA.
- A different developer, who has signed the CLA, could make a similar change.
- This PR could wait until after we have split all of the adapters out into separate GitHub repos (this is a change coming soonish as part of our [1.0 release](https://github.com/PixarAnimationStudios/OpenTimelineIO/projects/14)), at which time the owner of the kdenlive adapter plugin can accept this change under whatever terms they choose.

**Comment by scanlime** [comment]
You're aware that kdenlive import is broken right now, and the fix is quite straightforward? You probably already spent more time on the meeting than a fix would take.

**Comment by meshula** [comment]
@codingcatgirl indicated they're fine with anyone getting the patch into the codebase however that happens, so let's make that happen.

I've added a help wanted tag, and request that someone (1) with a signed CLA, and (2) familiar with kdenlive create a patch locally and test it, and submit a new PR with that patch, which we'll be happy to merge.

The purpose of testing locally is not doubt about whether @codingcatgirl made a mistake, I'm sure it's fine, but a simple check that the new PR is formed correctly.

The purpose of a new PR is to adhere to the legal framework surrounding submissions to this project, which says PRs need to originate with someone who has signed a contributor's agreement.

**Comment by gilou** [comment]
I tested it locally, and it works. I can submit the PR myself to help fix KDEnlive, but I'd let @codingcatgirl decide if she wants to move forward with the CLA as stated, as it's her fix, not mine.

I see the point of fixing stuff and making things happen, but there is no reason to be rude towards employees of a company that try to interact correctly with opensource communities. Yes, it's a little less convenient than just submitting a PR and flying away... But if their company decided that was the way to go to make that project opensource, maybe we can play by their rules, and avoid being harsh with fellow developers. There are (not too bad) reasons for CLAs to be required in these situations.

**Comment by jlskuz** [comment]
Thanks from the Kdenlive Team for patching this @codingcatgirl ! Pinging @vpinon (original author of the adapter) to might help here.

Looking forward to 1.0 and when the adapter can be maintained by the Kdenlive Team directly…

**Comment by apetrynet** [comment]
Hi @jlskuz!

> Looking forward to 1.0 when the adapter can be maintained by the Kdenlive Team directly…

As been discussed in a few TSC meetings and I believe you're referring to here, there is a plan to move contrib adapters out of the main OTIO repo and into individual repos/packages. This way maintainers of the adapters are not held back from updating the adapters outside of OTIO releases. Likewise, rarely maintained adapters won't hold back future releases of OTIO. (win win)

If the author(s) (@vpinon) of the kdenlive adapter agrees there shouldn't be anything in the way of starting the process of moving the adapter out of the core OTIO repo now before 1.0 is out. 
There is a template repo available [here](https://github.com/OpenTimelineIO/otio-plugin-template) to help with cases like this.

Please correct me if I'm wrong @jminor, but if you choose to create a separate kdenlive adapter repo, I would assume a PR removing the adapter from the main OTIO should be created before the standalone adapter hits pypi (or wherever) to avoid conflicting adapters.

**Comment by jlskuz** [comment]
> Hi @jlskuz!

Hi @apetrynet !
> 
> > Looking forward to 1.0 when the adapter can be maintained by the Kdenlive Team directly…
> 
> As been discussed in a few TSC meetings and I believe you're referring to here, there is a plan to move contrib adapters out of the main OTIO repo and into individual repos/packages. This way maintainers of the adapters are not held back from updating the adapters outside of OTIO releases. Likewise, rarely maintained adapters won't hold back future releases of OTIO. (win win)

I refer to @jminor's previous comment: https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/1160#issuecomment-1006983226

> If the author(s) (@vpinon) of the kdenlive adapter agrees there shouldn't be anything in the way of starting the process of moving the adapter out of the core OTIO repo now before 1.0 is out. There is a template repo available [here](https://github.com/OpenTimelineIO/otio-plugin-template) to help with cases like this.

I will check this with @vpinon (he's also part of the Kdenlive Team). But one important question: can, should or must the new repo live under the OpenTimelineIO name space on GitHub or can it/should it live on another namespace?

**Comment by apetrynet** [comment]
The idea behind the template is that anyone can use it as a starting point for writing plugins. Once "cloned" there are no stings attached and the repo lives under your namespace. The template only suggests a naming convention and so on to try unify plugins in the wild. There is nothing holding you from creating your adapter repo from scratch anywhere. 

One thing to consider if you choose a different license is what @meshula mentioned in #1185 about certain licenses potentially standing in the way of companies contributing to/using the adapter.

**Comment by vpinon** [comment]
Hello,
I don't have a strong opinion about moving kdenlive adapter to a separate repo...
For now I just slightly re-authored the patch (is this sufficient to get into the rules?), checked it passes 'make test && make lint' (on my system with python 3.9), hoping to unblock the situation for current users (PR #1191).
There is more work to do on kdenlive adapter, since the format has evolved to handle on-track transitions, but I haven't found time to work on this for a while...
Thanks @codingcatgirl and others taking care of otio!

**Comment by jminor** [comment]
Thanks @vpinon and @codingcatgirl for finding a path to fixing this!

**Pull request closed without merging**
